### PR TITLE
Fix gas computation and tx fee accumulation

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: process/block/shardblock.go [hunk 1]
```diff
@@ -260,6 +260,12 @@ func (sp *shardProcessor) ProcessBlock(
 		return err
 	}
 
+	//TODO: This part of code should be removed, it has been used only for debugging
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: These should be deleted

File: process/block/shardblock.go [hunk 2]
```diff
@@ -270,6 +276,19 @@ func (sp *shardProcessor) ProcessBlock(
...
```
Comment (SebastianMarian) [inline process/block/shardblock.go]: These should be deleted

File: process/smartContract/process.go [hunk 3]
```diff
@@ -149,6 +149,7 @@ func (sc *scProcessor) ExecuteSmartContractTransaction(
 
 	err := sc.processSCPayment(tx, acntSnd)
 	if err != nil {
+		log.Debug("process sc payment error", "error", err.Error())
```
Comment (SebastianMarian) [inline process/smartContract/process.go]: Maybe you can add more info about tx in these prints

File: process/smartContract/process.go [hunk 4]
```diff
@@ -201,7 +208,7 @@ func (sc *scProcessor) processIfError(
 	tx data.TransactionHandler,
 	returnCode string,
 ) error {
-	consumedFee := big.NewInt(0).SetUint64(tx.GetGasLimit() * tx.GetGasPrice())
+	consumedFee := big.NewInt(0).Mul(big.NewInt(0).SetUint64(tx.GetGasLimit()), big.NewInt(0).SetUint64(tx.GetGasPrice()))
```
Comment (SebastianMarian) [inline process/smartContract/process.go]: Good catch :+1:

File: process/smartContract/process.go [hunk 5]
```diff
@@ -222,7 +229,7 @@ func (sc *scProcessor) processIfError(
 
 	err = sc.scrForwarder.AddIntermediateTransactions(scrIfError)
 	if err != nil {
-		return nil
+		return err
```
Comment (SebastianMarian) [inline process/smartContract/process.go]: :+1:

File: process/block/preprocess/basePreProcess.go [hunk 6]
```diff
@@ -111,6 +112,7 @@ func (bpp *basePreProcess) saveTxsToStorage(
 		forBlock.mutTxsForBlock.RUnlock()
 
 		if txInfo == nil || txInfo.tx == nil {
+			log.Debug("missing transaction in saveTxsToStorage ", "type", dataUnit)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/basePreProcess.go]: done.

File: process/block/preprocess/gasComputation.go [hunk 7]
```diff
@@ -133,6 +132,7 @@ func (gc *gasComputation) ComputeGasConsumedByMiniBlock(
 	for _, txHash := range miniBlock.TxHashes {
 		txHandler, ok := mapHashTx[string(txHash)]
 		if !ok {
+			log.Debug("missing transaction in ComputeGasConsumedByMiniBlock ", "type", miniBlock.Type)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/gasComputation.go]: done

File: process/block/preprocess/gasComputation.go [hunk 8]
```diff
@@ -157,26 +157,21 @@ func (gc *gasComputation) ComputeGasConsumedByTx(
 	txReceiverShardId uint32,
 	txHandler data.TransactionHandler,
 ) (uint64, uint64, error) {
-
-	tx, ok := txHandler.(*transaction.Transaction)
-	if !ok {
-		return 0, 0, process.ErrWrongTypeAssertion
+	if check.IfNil(txHandler) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/gasComputation.go]: done.

File: process/block/preprocess/gasComputation.go [hunk 9]
```diff
@@ -157,26 +157,21 @@ func (gc *gasComputation) ComputeGasConsumedByTx(
...
+		return 0, 0, process.ErrNilTransaction
 	}
 
-	txGasLimitConsumption := gc.economicsFee.ComputeGasLimit(tx)
-	if tx.GasLimit < txGasLimitConsumption {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/gasComputation.go]: this is not used in interceptor. it is used only in basepreprocess and miniblocks compaction

File: process/block/preprocess/miniBlocksCompaction.go [hunk 10]
```diff
@@ -182,6 +182,7 @@ func (mbc *miniBlocksCompaction) expandMiniBlocks(miniBlocks block.MiniBlockSlic
 		for _, txHash := range miniBlock.TxHashes {
 			tx, ok := mbc.mapHashToTx[string(txHash)]
 			if !ok {
+				log.Debug("missing transaction in expandMiniBlocks ", "type", miniBlock.Type)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/miniBlocksCompaction.go]: done.

File: process/block/preprocess/miniBlocksCompaction.go [hunk 11]
```diff
@@ -255,6 +256,7 @@ func (mbc *miniBlocksCompaction) createMiniBlockForShard(miniBlock *block.MiniBl
 
 		tx, ok := mbc.mapHashToTx[string(txHash)]
 		if !ok {
+			log.Debug("missing transaction in createMiniBlockForShard ", "type", miniBlock.Type)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/miniBlocksCompaction.go]: done.

File: process/block/preprocess/rewardTxPreProcessor.go [hunk 12]
```diff
@@ -218,6 +218,7 @@ func (rtp *rewardTxPreprocessor) ProcessBlockTransactions(
 			txData := rtp.rewardTxsForBlock.txHashAndInfo[string(txHash)]
 			rtp.rewardTxsForBlock.mutTxsForBlock.RUnlock()
 			if txData == nil || txData.tx == nil {
+				log.Debug("missing transaction/missmatch in ProcessBlockTransactions ", "type", block.RewardsBlock)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/rewardTxPreProcessor.go]: done.

File: process/block/preprocess/smartContractResults.go [hunk 13]
```diff
@@ -216,6 +221,7 @@ func (scr *smartContractResults) ProcessBlockTransactions(
 			txInfo := scr.scrForBlock.txHashAndInfo[string(txHash)]
 			scr.scrForBlock.mutTxsForBlock.RUnlock()
 			if txInfo == nil || txInfo.tx == nil {
+				log.Debug("missing transaction in ProcessBlockTransactions ", "type", block.SmartContractResultBlock)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/smartContractResults.go]: done.

File: process/block/preprocess/transactions.go [hunk 14]
```diff
@@ -240,6 +239,7 @@ func (txs *transactions) ProcessBlockTransactions(
 			txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
 
 			if txInfo == nil || txInfo.tx == nil {
+				log.Debug("missing transaction in ProcessBlockTransactions ", "type", block.TxBlock)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/block/preprocess/transactions.go]: done.

**Pull request merged**